Feature/automated testing

BenBradnick opened the pull request:
This adds a ScanRunnerPart for testing a scan block. Scan parameters are first loaded from a user-written YAML file, and then the part can be ran to go through each set of parameters. The part writes scan data and a small report file inside a defined top-level output directory.

BenBradnick commented:
I combined your suggestion into a single object as the generator contains all of the physical parameters.

Let me know if there is anything else I should modify - otherwise I believe it is ready.

thomascobb commented on malcolm/modules/scanning/parts/scanrunnerpart.py:
```diff
@@ -0,0 +1,498 @@
+        except IOError:
+            self.set_runner_state(RunnerStates.FAULT)
+            self.runner_status_message.set_value("Could not read scan file")
+            raise IOError("Could not read scan file")
```
```suggestion
            raise
```
I would just raise here, as the error message from open() is probably more descriptive and the traceback is useful to have

BenBradnick replied:
I have changed this accordingly.

thomascobb commented on malcolm/modules/scanning/parts/scanrunnerpart.py:
```diff
@@ -0,0 +1,498 @@
+        except yaml.YAMLError:
+            self.set_runner_state(RunnerStates.FAULT)
+            self.runner_status_message.set_value("Could not parse scan file")
+            raise yaml.YAMLError("Could not parse scan file")
```
Possibly ditto? Would be useful to have the traceback

BenBradnick replied:
This has also been changed.

thomascobb commented on malcolm/modules/scanning/parts/scanrunnerpart.py:
```diff
@@ -0,0 +1,498 @@
+        generators = []
+        generators_dict = entry['generators']
+        for generator in generators_dict:
+            generators.append(self.parse_line_generator(generator['line']))
```
```suggestion
            generators.append(LineGenerator.from_dict(generator['line']))
```
Should work...

BenBradnick replied:
It did work, so I've committed the change.

thomascobb commented on malcolm/modules/scanning/parts/scanrunnerpart.py:
```diff
@@ -0,0 +1,498 @@
+        return sub_directory
+
+    @add_call_types
+    def run(self, context):
```
I take it we don't care about aborting these scans when they are kicked off?

BenBradnick replied:
I was planning on adding it at some point. What would the best way of handling this be?

thomascobb replied:
Store context from run, then make a new function abort() that calls context.stop(). That will catch run() when it is waiting for the scan to run, but not abort the current running scan. To do that you would also have to abort() the last scan that was run.

thomascobb commented on malcolm/modules/scanning/parts/scanrunnerpart.py:
```diff
@@ -0,0 +1,498 @@
+
+        # Check if scan can be reset or run
+        while scan_block.state.value is RunnableStates.ABORTING:
+            cothread.Sleep(0.1)
```
Would be better to do context.sleep

BenBradnick replied:
This is now using context.sleep.

thomascobb commented on malcolm/modules/scanning/parts/scanrunnerpart.py:
```diff
@@ -0,0 +1,487 @@
+        kwargs_list = ['continuous', 'delay_after']
+        kwargs = self.get_kwargs_from_dict(entry, kwargs_list)
+
+        return CompoundGenerator(generators, duration=duration, **kwargs)
```
You should be able to do CompoundGenerator.from_dict(entry) here too, as long you have set entry["generators"] = generators

BenBradnick replied:
I have made this change and added a check for a negative generator duration.

BenBradnick commented:
I have added abort which first stops the stored context and then aborts the current scan (I believe this is the correct order?).

More tests were also added to improve coverage.

BenBradnick commented:
I believe this is now at the point ready for merging, if you are happy @thomascobb
